### PR TITLE
Fix source position of templates with empty body/parents

### PR DESCRIPTION
**Pull request by som-snytt**
The position of those absent elements was incorrectly taken as following on the next line.

Fixes scala/bug#12470

### CODE REVIEW

**Review by dwijnand — APPROVED** [review]
Looks reasonable to me.  The CI is failing on checkfile updates of positions.  Didn't review those, though.

**Comment by som-snytt** [comment]
The checkfile updates are constructor and class parents repositioned from after template to start of class.

It would be interesting to have a test that compares dotty classfile output.

File: src/compiler/scala/tools/nsc/ast/parser/Parsers.scala [hunk 1]
```diff
@@ -3246,8 +3246,8 @@ self =>
         // regarding AnyVal constructor in AddInterfaces.
         DefDef(NoMods, nme.CONSTRUCTOR, Nil, ListOfNil, TypeTree(), Block(Nil, literalUnit))
       )
-      val parentPos = o2p(in.offset)
       val tstart1 = if (body.isEmpty && in.lastOffset < tstart) in.lastOffset else tstart
+      val parentPos = if (parents.isEmpty) o2p(tstart1) else o2p(in.offset)
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/ast/parser/Parsers.scala]: Was `in.offset` really ever correct here? It's the position after the template, why would syntethic parents use that positoin? Note that after your fix, `o2p(in.offset)` is never used; within `gen.mkParents` we the only use of that position is `if (parents.isEmpty) atPos(parentPos)(scalaAnyRefConstr) :: Nil`.

File: src/compiler/scala/tools/nsc/ast/parser/Parsers.scala [hunk 2]
```diff
@@ -3246,8 +3246,8 @@ self =>
         // regarding AnyVal constructor in AddInterfaces.
         DefDef(NoMods, nme.CONSTRUCTOR, Nil, ListOfNil, TypeTree(), Block(Nil, literalUnit))
       )
-      val parentPos = o2p(in.offset)
       val tstart1 = if (body.isEmpty && in.lastOffset < tstart) in.lastOffset else tstart
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/ast/parser/Parsers.scala]: I got it after a while. `tstart` is the start of the token after the template name, so for a simple `class Plain; [rest]` it points to the start of `[rest]`. This fix makes it point to the end of the name `Plain`.

Maybe add a comment to the source? And/or rename `tstart`.

**Review by lrytz — COMMENTED** [review]
LGTM, but some nitpicking.. Thanks for the fix!

File: src/compiler/scala/tools/nsc/ast/parser/Parsers.scala (hunk 1 above)
Comment (som-snytt) [reply src/compiler/scala/tools/nsc/ast/parser/Parsers.scala]: I did not have capacity to overthink it at the time of fix, but I'll try to give it more clarity.

In particular, I had started at backend and bytecode testing, so by the time I arrived at position assignment in frontend, I was looking for closure. However, it is all worthwhile.

**Comment by som-snytt** [comment]
I noticed that `tstart` in the `class` case is at the type params, and that scalac ~incorrectly~ accepts:
```
class K
[T]
```
which dotty rejects.

This was intentional:
```
newLineOptWhenFollowedBy(LBRACKET)
```
Edit: it doesn't see a NL here. At `[`, a NL is not inserted by scanner because `[` cannot start a statement.

Dotty has `canStartExprTokens3` which includes `openParensTokens` which includes `LBRACKET`. ~That is probably just a bug.~ Spec 1.2 on NL specifically excludes LBRACKET. Actually it's to accommodate new type lamba syntax.

**Comment by som-snytt** [comment]
The old code comment before Miles' "switch over to new scanner/parser":
```
SUPERTYPE | HASH | RPAREN | RBRACKET | RBRACE => // todo: add LBRACKET
```
So the spurious `newLineOptWhenFollowedBy(LBRACKET)` due to Martin's "small fixes to newline behavior" was under the old regime, before `[` was added to exclusions of what can begin a statement.

**Comment by SethTisue** [comment]
I'll merge this as soon as it passes Jenkins, in order to have a new release candidate, but I won't be building 2.13.7 itself before Monday, so there's still some time for additional review, second thoughts, etc.

**Comment by som-snytt** [comment]
Oops, I paused for dinner with the family. My second thought is at https://github.com/scala/scala/pull/9796 which deletes the obsolete line mentioned above.

**Pull request merged**